### ISSUE TITLE
Viewport: scaled + clampedToEntities prevents mouseLooking full scene

### ISSUE THREAD

**Issue by naasking** [craftyjs/Crafty#1010]
Take a look at: https://jsfiddle.net/0btLn8ep/

So if you click and move, it centers all elements like described in my last ticket. Now if you scroll/zoom in until the yellow box is a little out of frame on the bottom and right sides, you'll see that you can't drag the viewport to see the full yellow square.

**Comment by 200sc** [comment]
If your concern is that the zoom doesn't modify the viewport's coordinates it uses for clampToEntities appropriately, you should either (1) roll your own area the viewport can't exist in or (2) not use viewport zooming/scaling, as it's unstable in that respect.

As far as rolling your own viewport, I'd recommend using Crafty.viewport.bounds. I've got an open source project that uses it with Crafty.viewport.follow, [here](https://bitbucket.org/StephenPatrick/entry-for-ld34/src/b1ca1aba705b3588a3c8f510c0f9d3c48f8b1afd/src/game.js?at=master&fileviewer=file-view-default#game.js-80).

**Comment by naasking** [comment]
clampToEntities seems to have no effect on the fact that you can't scroll to fully view off-screen objects when you're zoomed in. I've updated the fiddle with some text on the yellow box to show just how much is cut off:

https://jsfiddle.net/0btLn8ep/1/

Scrolling with mouselook often gets stuck on the 5 or 6, but you can never see 7 even though it's clearly within the box.

The docs don't list any scaling limitations on the default viewport, If it's an inherent limitation of the default viewport that won't get fixed, then I guess this bug is merely a documentation issue.

**Comment by 200sc** [comment]
Try [Crafty.viewport.clampToEntities](https://jsfiddle.net/0btLn8ep/2/), not Crafty.clampToEntities.

(As far as scaling goes, I've just found it very difficult to work with scaling / zooming while having other components or otherwise trying to modify the viewport. It has been easier to double the size of all of my sprites than to zoom the viewport 2x)

**Comment by naasking** [comment]
You are totally correct. I've clearly been starting at code for too long if I missed that one too.

That certainly addresses what I need, but the behaviour when clampToEntities is true still seems incorrect. Or am I totally off base here in thinking you should be able to see the full object?

**Comment by 200sc** [comment]
You should be. The way that Crafty scales and zooms doesn't work well with other viewport functions, though. A proper investigation into what the disconnect is between how Crafty scales and what other viewport-affecting functions want would be great. I imagine there are just hanging references to literal viewport positions, unaffected by the viewport's zoom level, that need to be replaced. For the time being, for functionality, I'd suggest using viewport.bounds or avoiding scale / zoom.

As an example, SolidHitBox has never worked nicely for me with viewport scaling.
